File: Java/src/irrgarten/Game.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;

/**
 *
 * @author javie
 */
public class Game {
    private static final int MAX_ROUNDS = 10;
    private int currentPlayerIndex;
    private String log;
    
    private Labyrinth labyrinth;
    private ArrayList<Player> players;
    private ArrayList<Monster> monsters;
    private Player currentPlayer;    

    
    /**
     * Constructor de la clase Game que inicializa el juego con un número dado de jugadores.
     * Crea un laberinto, jugadores y monstruos aleatorios, y configura el estado inicial del juego.
     * 
     * @param nplayers Número de jugadores que participarán en el juego.
     */
    public Game(int nplayers) {
        
        log = "";
        labyrinth = new Labyrinth(4, 4, 3, 3);
        players = new ArrayList<>();
        monsters = new ArrayList<>();

        for (int i = 0; i < nplayers; i++) {
            char number = (char)('0' + i);
            Player player = new Player(number, Dice.randomIntelligence(), Dice.randomStrength());
            players.add(player);
        }
        
        configureLabyrinth();
        currentPlayerIndex = Dice.whoStars(nplayers); 
        currentPlayer = players.get(currentPlayerIndex);
              
        labyrinth.spreadPlayers(players);
    }
    
    public boolean finished(){
        return labyrinth.haveAWinner();
    }
    
    
    /**
     * Realiza el siguiente paso del juego para el jugador actual.
     * Maneja el movimiento del jugador, los combates con monstruos y la gestión de recompensas y resurrecciones.
     * 
     * @param preferredDirection La dirección preferida para que el jugador se mueva.
     * @return True si el juego ha terminado después del paso, False si no.
     */
    public boolean nextStep(Directions preferredDirection){
        String log = "";
        boolean dead = currentPlayer.dead();
        if(!dead){
            Directions direction = actualDirection(preferredDirection);
            if(direction != preferredDirection){
                logPlayerNoOrders();
            }
            Monster monster = labyrinth.putPlayer(direction, currentPlayer);
            
            if(monster == null){
                logNoMonster();
            } else {
                GameCharacter winner = combat(monster);
                manageReward(winner);
            }
        } else {
            manageResurrection();
        }
        
        boolean endGame = finished();
        
        if(!endGame){
            nextPlayer();
        }
        
        return endGame;
    }
    
    /**
     * Obtiene el estado actual del juego incluyendo el estado del laberinto, jugadores y monstruos.
     * 
     * @return El estado actual del juego encapsulado en un objeto GameState.
     */
    public GameState getGameState(){
        String labyrinthState = labyrinth.toString();
        String playerState = "";
        for (int i = 0; i < players.size(); i++) {
            Player player = players.get(i);
            playerState += player.toString() + "\n";
        }
        String monsterState = "";
        for (int i = 0; i < monsters.size(); i++) {
            Monster monster = monsters.get(i);
            monsterState += monster.toString() + "\n";
        }
        return new GameState(labyrinthState, playerState, monsterState, currentPlayerIndex, finished(), log);
    }
    
    private void configureLabyrinth(){
        labyrinth.addBlock(Orientation.VERTICAL, 3, 2, 1);
        labyrinth.addBlock(Orientation.HORIZONTAL, 1, 2, 2);
        String nombre1 = "M1";
        String nombre2 = "M2";
        
        Monster monster1 = new Monster(nombre1, Dice.randomIntelligence(), 
                            Dice.randomStrength());
        Monster monster2 = new Monster(nombre2, Dice.randomIntelligence(), 
                            Dice.randomStrength());
    
        labyrinth.addMonster(Dice.randomPos(4), Dice.randomPos(4), monster1);
        labyrinth.addMonster(Dice.randomPos(4), Dice.randomPos(4), monster2);
        
        monsters.add(monster1);
        monsters.add(monster2);
    }
    
    private void nextPlayer(){
        if(currentPlayerIndex < players.size() -1){
            currentPlayerIndex += 1;
            currentPlayer = players.get(currentPlayerIndex);
        } else {
            currentPlayerIndex = 0;
            currentPlayer = players.get(currentPlayerIndex);
        }
    }
    
    
    Directions actualDirection(Directions preferredDirection){
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();
        ArrayList<Directions> validMoves = new ArrayList<>();
        validMoves = labyrinth.validMoves(currentRow, currentCol);
        Directions output = currentPlayer.move(preferredDirection, validMoves);
        return output;
    }
    
    private GameCharacter combat(Monster monster){
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        float playerAttack = currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        
        while(!lose && rounds < MAX_ROUNDS){
            winner = GameCharacter.MONSTER;
            rounds++;
            float monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);
            if(!lose){
                playerAttack = currentPlayer.attack();
                lose = monster.defend(playerAttack);
                winner = GameCharacter.PLAYER;
            }
        }
        
        logRounds(rounds, MAX_ROUNDS);
        return winner;
    }
    
    private void manageReward(GameCharacter winner){
        if(winner == GameCharacter.PLAYER){
            currentPlayer.receiveReward();
            logPlayerWon();
        } else {
            logMonsterWon();
        }
    }
    
    private void manageResurrection(){
        boolean resurrect = Dice.resurrectPlayer();
        
        if(resurrect){
            currentPlayer.resurrect();
            logResurrected();
        } else {
            logPlayerSkipTurn();
        }
    }
    
    private void logPlayerWon() {
        log = "Player " + currentPlayerIndex + " has won the combat.\n";
    }

    private void logMonsterWon() {
        log = "The monster has won the combat.\n";
    }

    private void logResurrected() {
        log = "Player " + currentPlayerIndex + " has resurrected.\n";
    }

    private void logPlayerSkipTurn() {
        log = "Player " + currentPlayerIndex + " has lost the turn for being dead.\n";
    }

    private void logPlayerNoOrders() {
        log = "Player " + currentPlayerIndex + " do not follow the human instructions (not posible).\n";
    }

    private void logNoMonster() {
        log = "Player " + currentPlayerIndex + " has moved to an empty cell or was unable to move.\n";
    }

    private void logRounds(int rounds, int max) {
        log = rounds + " rounds of " + max + " passed.\n";
    }
}

File: Ruby/game.rb
# frozen_string_literal: true
require_relative 'orientation'
require_relative 'labyrinth'
require_relative 'player'
require_relative 'dice'
require_relative 'game_character'
require_relative 'labyrinth'
require_relative 'game_state'
require_relative 'monster'

module Irrgarten
  class Game

    @@MAX_ROUNDS = 10

    def initialize(nplayers)
      @log = ''
      @players = []
      @monsters = []

      for i in 1..nplayers
        number = i.to_s
        player = Player.new(number, Dice.random_intelligence, Dice.random_strength)
        @players << player
      end

      @current_player_index = Dice.who_starts(nplayers)
      @current_player = @players[@current_player_index]

      @labyrinth = Labyrinth.new(4, 4, 3, 3)
      configure_labyrinth
      @labyrinth.spread_players(@players)
    end

    def finished
      @labyrinth.have_a_winner
    end

    def next_step(preferred_direction)
      log = ''
      dead = @current_player.dead

      if !dead
        direction = actual_direction(preferred_direction)
        if direction != preferred_direction
          log_player_no_orders
        end
        monster = @labyrinth.put_player(direction, @current_player)
        if monster == nil
          log_no_monster
        else
          winner = combat(monster)
          manage_reward(winner)
        end
      else
        manage_resurrection
      end
      end_game = finished
      if !end_game
        next_player
      end
      end_game
    end

    def game_state
      labyrinth_state = @labyrinth.to_s
      player_state = ''
      for player in @players
        player_state += player.to_s
      end

      monster_state = ''
      for monster in @monsters
        monster_state += monster.to_s
      end

      game_state = GameState.new(labyrinth_state, player_state, monster_state, @current_player.number, finished, @log)
    end

    def configure_labyrinth
      @labyrinth.add_block(Orientation::VERTICAL, 3, 2, 1)
      @labyrinth.add_block(Orientation::HORIZONTAL, 1, 2, 2)

      monster1 = Monster.new('M1', Dice.random_intelligence, Dice.random_strength)
      monster2 = Monster.new('M2', Dice.random_intelligence, Dice.random_strength)

      @labyrinth.add_monster(Dice.random_pos(4), Dice.random_pos(4), monster1)
      @labyrinth.add_monster(Dice.random_pos(4), Dice.random_pos(4), monster2)
      @monsters << monster1
      @monsters << monster2
    end

    def next_player
      if @current_player_index < @players.size - 1
        @current_player_index += 1
        @current_player = @players[@current_player_index]
      else
        @current_player_index = 0
        @current_player = @players[@current_player_index]
      end
    end

    def actual_direction(preferred_direction)
      current_row = @current_player.row
      current_col = @current_player.col
      valid_moves = []
      valid_moves = @labyrinth.valid_moves(current_row, current_col)
      output = @current_player.move(preferred_direction, valid_moves)
    end

    def combat(monster)
      rounds = 0
      winner = GameCharacter::PLAYER
      player_attack = @current_player.attack
      lose = monster.defend(player_attack)

      while !lose && rounds < @@MAX_ROUNDS
        winner = GameCharacter::MONSTER
        rounds += 1
        monster_attack = monster.attack
        lose = @current_player.defend(monster_attack)
        if !lose
          player_attack = @current_player.attack
          lose = monster.defend(player_attack)
          winner = GameCharacter::PLAYER
        end
      end
      log_rounds(rounds, @@MAX_ROUNDS)
      winner
    end

    def manage_resurrection
      resurrect = Dice.resurrect_player

      if resurrect
        @current_player.resurrect
        log_resurrected
      else
        log_player_skip_turn
      end
    end

    def manage_reward(winner)
      if winner == GameCharacter::PLAYER
        @current_player.receive_reward
        log_player_won
      else
        log_monster_won
      end
    end

    def log_player_won
      @log = "The player " + @current_player_index.to_s + " has won the combat.\n"
    end

    def log_monster_won
      @log = "The monster has won the combat.\n"
    end

    def log_resurrected
      @log = "The player " + @current_player_index.to_s + " has resurrected.\n"
    end

    def log_player_skip_turn
      @log = "The player " + @current_player_index.to_s + " has lost turn for being dead.\n"
    end

    def log_player_no_orders
      @log = "The player " + @current_player_index.to_s + " did not follow the human player's instructions (not possible).\n"
    end

    def log_no_monster
      @log = "The player " + @current_player_index.to_s + " moved to an empty cell or could not move at all.\n"
    end

    def log_rounds(rounds, max)
      @log = "There have been #{rounds} of #{max} rounds of combat.\n"
    end
  end
end